NorthstarDLL/masterserver.cpp

```diff
@@ -814,595 +692,385 @@ void MasterServerManager::AuthenticateWithServer(char* uid, char* playerToken, c
-										escapedPlaylistNew,
-										g_ServerAuthenticationManager->m_additionalPlayerData.size(),
-										maxPlayers,
-										escapedPasswordNew)
-										.c_str());
-
-								curl_free(escapedNameNew);
-								curl_free(escapedDescNew);
-								curl_free(escapedMapNew);
-								curl_free(escapedPlaylistNew);
-								curl_free(escapedPasswordNew);
+								spdlog::error(
+									"Failed reading masterserver authentication response: encountered parse error \"{}\"",
+									rapidjson::GetParseError_En(serverAddedJson.GetParseError()));
+								goto REQUEST_END_CLEANUP;
 							}
 
-							curl_mime* mime = curl_mime_init(curl);
-							curl_mimepart* part = curl_mime_addpart(mime);
-
-							curl_mime_data(part, m_sOwnModInfoJson.c_str(), m_sOwnModInfoJson.size());
-							curl_mime_name(part, "modinfo");
-							curl_mime_filename(part, "modinfo.json");
-							curl_mime_type(part, "application/json");
-
-							curl_easy_setopt(curl, CURLOPT_MIMEPOST, mime);
-
-							CURLcode result = curl_easy_perform(curl);
-
-							// defensive check, as m_sOwnServerId could be set to null before this request gets processed
-							if (!*m_sOwnServerId)
-								return;
-
-							if (result == CURLcode::CURLE_OK)
+							if (!serverAddedJson.IsObject())
 							{
-								rapidjson_document serverAddedJson;
-								serverAddedJson.Parse(readBuffer.c_str());
-
-								if (!serverAddedJson.HasParseError() && serverAddedJson.IsObject())
-								{
-									if (serverAddedJson.HasMember("id") && serverAddedJson["id"].IsString())
-									{
-										strncpy(m_sOwnServerId, serverAddedJson["id"].GetString(), sizeof(m_sOwnServerId));
-										m_sOwnServerId[sizeof(m_sOwnServerId) - 1] = 0;
-									}
-
-									if (serverAddedJson.HasMember("serverAuthToken") && serverAddedJson["serverAuthToken"].IsString())
-									{
-										strncpy(
-											m_sOwnServerAuthToken,
-											serverAddedJson["serverAuthToken"].GetString(),
-											sizeof(m_sOwnServerAuthToken));
-										m_sOwnServerAuthToken[sizeof(m_sOwnServerAuthToken) - 1] = 0;
-									}
-								}
+								spdlog::error("Failed reading masterserver authentication response: root object is not an object");
+								goto REQUEST_END_CLEANUP;
 							}
-							else
-								spdlog::warn("Heartbeat failed with error {}", curl_easy_strerror(result));
 
-							curl_easy_cleanup(curl);
-							Sleep(10000);
-						} while (*m_sOwnServerId);
-					});
+							if (serverAddedJson.HasMember("error"))
+							{
+								spdlog::error("Failed reading masterserver response: got fastify error response");
+								spdlog::error(readBuffer);
 
-				heartbeatThread.detach();
-			}
-			else
-			{
-				spdlog::error("Failed adding self to server list: error {}", curl_easy_strerror(result));
-				m_bSuccessfullyConnected = false;
-			}
+								if (serverAddedJson["error"].HasMember("enum") &&
+									!strcmp(serverAddedJson["error"]["enum"].GetString(), "DUPLICATE_SERVER"))
```

Comment:
As @RoyalBlue1 mentioned, it retries connection even when not `DUPLICATE_SERVER`. Not sure why this happens exactly though. ^^

Behaviour can be easily observed by just starting game, going into `mp_lobby` and watching console

Reply:
Seems to be fixed somewhere between opening issue and b03070d5b4575581ff6183cada8083446caa351e